### PR TITLE
client: added browser tests

### PR DESCRIPTION
**Pull request by Bullrich**
To be merged **after** #240 

Added playwright tests to the project.
Now it run tests on a headless browser verifying that the implementation works correctly.

It sets the public captcha testing key which always approves the operation (thanks @mordamax for that tip).

**9** tests were added to the project.

### CODE REVIEW

**Review by altaua — APPROVED** [review]

**Review by coderobe — APPROVED** [review]

File: client/tests/test.ts [hunk 1]
```diff
@@ -1,6 +1,168 @@
...
+
+const getFormElements = async (page: Page, getCaptcha = false) => {
+	let captcha: Locator = {} as Locator;
+	if (getCaptcha) {
+		// ?: Hack. We need to wait for the frame to load and then invade it.
+		await page.reload();
+		const captchaFrame = await new Promise<Frame>(function (resolve, reject) {
+			let i = 0;
+			// function that waits for the frame and timeouts after 3 seconds
+			(function waitForFrame() {
```
Comment (mutantcornholio) [inline client/tests/test.ts:19]: `import { until } from "opstooling-js";`?

File: client/tests/test.ts [hunk 2]
```diff
@@ -1,6 +1,168 @@
...
+		});
+		captcha = captchaFrame?.locator("#recaptcha-anchor") as Locator;
+	}
+	return {
+		address: page.getByPlaceholder("Enter your address"),
+		useParachain: page.getByLabel("Use parachain"),
+		// value changes after useParachain is checked
+		parachain: page.getByPlaceholder("Using Relay chain"),
+		captcha,
+		submit: page.getByRole("button", { name: "Submit" })
```
Comment (mutantcornholio) [inline client/tests/test.ts:42]: I'd recommend adding special markers (classname, data-attr, make your pick) to the elements in code, to search for them later in tests. 

"Element not found" errors are the nastiest.

File: client/tests/test.ts (hunk 1 above)
Comment (Bullrich) [reply client/tests/test.ts:24]: I'm having a problem while importing the package. It is empty.
Have you encounter the same problem?

image

File: client/tests/test.ts (hunk 2 above)
Comment (rzadp) [reply client/tests/test.ts:47]: I second that.

**Review by rzadp — COMMENTED** [review]
Have you considered testing against [multiple browsers](https://playwright.dev/docs/test-configuration#multiple-browsers)?

File: client/tests/test.ts (hunk 2 above)
Comment (Bullrich) [reply client/tests/test.ts]: I had an issue while setting it up but now I fixed it. Thank you for your recommendation. Now the code is cleaner!

File: client/tests/test.ts [hunk 3]
```diff
@@ -1,6 +1,193 @@
...
+			if (req.url() === url) {
+				const data = req.postDataJSON() as FormSubmit;
+				expect(data.address).toEqual(myAddress);
+				return !!data.recaptcha;
+			}
+			return false;
+		});
+		await submit.click();
+		// verify that the post request is correct
+		await request;
```
[the remark this answers is not in the document]
Comment (Bullrich) [reply client/tests/test.ts:141]: Oh! I need to add error handling! Well done in reminding me!

File: client/tests/test.ts (hunk 1 above)
Comment (Bullrich) [reply client/tests/test.ts:24]: Thanks @rzadp, it was imported now.

But now I'm having problems with playwright importing it:
> Error: No "exports" main defined in /Users/asd/Repositories/substrate-matrix-faucet/client/node_modules/opstooling-js/package.json imported from /Users/ads/Repositories/substrate-matrix-faucet/client/tests/test.ts

To be honest I don't wanna try to fight the compiler to only use one function, is it ok if I just copy and paste the method?

Comment (mutantcornholio) [reply client/tests/test.ts:24]: I guess, let's just leave it like that for now, and revisit it later.

**Review by mutantcornholio — APPROVED** [review]
LGTM :)

**Comment by Bullrich** [comment]
> Have you considered testing against [multiple browsers](https://playwright.dev/docs/test-configuration#multiple-browsers)?

I tried to but it seems to fail with github actions. It's not good if we can not automate it.

**Pull request merged**